integration_tests/e2e/confirm-achieve-goal.cy.ts

```diff
@@ -0,0 +1,152 @@
+        })
+
+        planOverview.agreePlan()
+      })
+    })
+
+    it('Displays agree plan page correctly on load', () => {
+      cy.get<Goal>('@newGoal').then(goal => {
+        cy.visit(`/confirm-if-achieved/${goal.uuid}`)
+        cy.get('h1').contains('has achieved this goal')
+        cy.get('.govuk-summary-card__title').should('contain', goal.title)
+      })
+      cy.checkAccessibility()
+    })
+  })
+
+  describe('Submission behaviour', () => {
+    beforeEach(() => {
+      cy.get<PlanType>('@plan').then(plan => {
+        cy.addGoalToPlan(plan.uuid, DataGenerator.generateGoal({})).then(goal => {
+          cy.wrap(goal).as('newGoal')
+          cy.addStepToGoal(goal.uuid, DataGenerator.generateStep({}))
+        })
+
+        planOverview.agreePlan()
+        cy.get<Goal>('@newGoal').then(goal => {
+          cy.visit(`/update-goal-steps/${goal.uuid}`)
+
+          cy.get('#step-status-1').select('Completed') // select completed status
+
+          cy.get('button').contains('Save goal and steps').click()
+        })
+      })
+    })
+
+    it('Confirm goal achieved successfully without optional note', () => {
+      achieveGoal.isGoalAchievedRadio('yes')
+      cy.get('button').contains('Save and continue').click()
+      cy.url().should('include', 'plan?type=achieved&status=achieved')
+      cy.get(':nth-child(3) > .moj-sub-navigation__link')
+      cy.get('.moj-sub-navigation__link').eq(2).should('contain', 'Achieved goals (1)')
+      cy.get('.govuk-summary-card').should('contain', 'Marked as achieved on')
+
+      cy.get<Goal>('@newGoal').then(goal => {
+        cy.visit(`/view-achieved-goal/${goal.uuid}`)
+      })
+      cy.get('.govuk-body').should('contain', 'Marked as achieved on')
+      cy.checkAccessibility()
+    })
+
+    it('Confirm goal achieved successfully with optional note', () => {
+      achieveGoal.isGoalAchievedRadio('yes')
+      cy.get('#goal-achievement-helped').type('Some optional text in the achievement note field')
+      cy.get('button').contains('Save and continue').click()
+      cy.url().should('include', 'plan?type=achieved&status=achieved')
+      cy.get(':nth-child(3) > .moj-sub-navigation__link')
+      cy.get('.moj-sub-navigation__link').eq(2).should('contain', 'Achieved goals (1)')
+
+      cy.get('.moj-sub-navigation__link').eq(2).click()
+      planOverview.getSummaryCard(0).get('a').contains('View details').click()
+      cy.get('.govuk-details__text').should('contain', 'Some optional text in the achievement note field')
+      cy.get('p.govuk-body').should('contain', 'Marked as achieved on ')
+      cy.get('a.govuk-back-link').should('have.attr', 'href').and('include', '/plan?type=achieved')
+      cy.get('.govuk-button--secondary').should('have.length', 0)
+      cy.checkAccessibility()
+    })
+
+    it('Confirm errors are displayed with optional note of more than 4000 characters', () => {
+      const lorem = faker.lorem.paragraphs(40)
```

Comment:
over in add-steps.cy.ts we have a function called `generateStringOfLength` - would we be better off extracting that into a utils class we can call in cases like this? It would be useful to be able to test with a string of a known length I think.

Reply:
added integrationUtils, yeah it's definitely useful and easier